### PR TITLE
Include the .po files from Khan Academy's Crowdin during .mo generation.

### LINKED ISSUE #225: Include the .po files from Khan Academy's Crowdin account during `update_language_packs`.
**Branch** master

**The problem**
Approved translations from Khan Academy's Crowdin account are not included in the language pack when `update_language_packs` is ran.

To reproduce:
* on central, run `python manage.py update_language_packs -l de --no-dubbed --no-srts`
* notice that there are a lot of messages ignoring the .po file like this
> DEBUG:kalite:Ignoring /var/folders/40/5c3y6nhs7tv00vv1nq5g0nvm0000gn/T/tmpJ1FTAW/4_low_priority/learn.partner-content.tate.exercises-de.po because it's NOT for version 0.13.
D

**Possible reason**
This is because at the `produce_outputs` function, we ignore the `.po` files which do not belong to the same `SHORTVERSION`.  Since there's no version mark for the KA Crowdin .po files, they are ignored by the function, thus we don't see the approved translations at Perseus exercises on distributed.

**Possible solution**
Include the `.po` files from Khan Academy's Crowdin during `.mo` generation.  DO NOT exclude them based on `SHORTVERSION`.

### PR DESCRIPTION
**Pull request by arceduardvincent**
Hi @aronasorman and @cpauya - this fixes #225 

I include the `.po` files during the generation of `.mo` files and NOT exclude them based on SHORTVERSION.

### CODE REVIEW

File: centralserver/i18n/management/commands/update_language_packs.py [hunk 1]
```diff
@@ -560,13 +560,12 @@ def produce_outputs(src_po_files, dest_path, lang_code):
                 # Make sure we only concatenate .po files of the same version that we need.
                 versioned_po_filename = os.path.join("versioned", "%s-django") % (version,)
                 kalite_po_filename = os.path.join("KA Lite UI", "kalite-%s.po") % (lang_code,)
+                src_po = polib.pofile(src_file)
+                build_po.merge(src_po)
+                #logging if the files is in the version
                 if versioned_po_filename in src_file or kalite_po_filename in src_file:
```
Comment (aronasorman) [inline centralserver/i18n/management/commands/update_language_packs.py]: Maybe just remove this `if` statement altogether?

Comment (arceduardvincent) [inline centralserver/i18n/management/commands/update_language_packs.py]: @aronasorman yeah I will, I am thinking to delete that one.

File: centralserver/i18n/management/commands/update_language_packs.py [hunk 2]
```diff
@@ -558,15 +558,9 @@ def produce_outputs(src_po_files, dest_path, lang_code):
...
-                versioned_po_filename = os.path.join("versioned", "%s-django") % (version,)
-                kalite_po_filename = os.path.join("KA Lite UI", "kalite-%s.po") % (lang_code,)
-                if versioned_po_filename in src_file or kalite_po_filename in src_file:
-                    logging.debug('Concatenating %s with %s...' % (src_file, build_file))
-                    src_po = polib.pofile(src_file)
-                    build_po.merge(src_po)
-                else:
-                    logging.debug("Ignoring %s because it's NOT for version %s." %
-                                  (src_file, version,))
+                src_po = polib.pofile(src_file)
```
Comment (cpauya) [inline centralserver/i18n/management/commands/update_language_packs.py]: Please don't just remove the `if` statement here because we use it to build versioned .mo files as per #210.

There must be a way to preserve that functionality while fixing this issue.

Comment (arceduardvincent) [inline centralserver/i18n/management/commands/update_language_packs.py]: If the .po files are existed in the `src_file ` then we will execute the method `build_po.merge(src_po)` it will merge/build the po files.

File: centralserver/i18n/management/commands/update_language_packs.py [hunk 3]
```diff
@@ -558,15 +558,22 @@ def produce_outputs(src_po_files, dest_path, lang_code):
...
-                    logging.debug('Concatenating %s with %s...' % (src_file, build_file))
-                    src_po = polib.pofile(src_file)
+                    logging.debug("Concatenating %s with %s..." % (src_file, build_file))
+                    src_po = polib.pofile(
+                        src_file)
                     build_po.merge(src_po)
                 else:
-                    logging.debug("Ignoring %s because it's NOT for version %s." %
-                                  (src_file, version,))
+                    if os.path.basename(src_file):
```
Comment (cpauya) [inline centralserver/i18n/management/commands/update_language_packs.py]: This will catch all `.po` files passed to it, including other versions of ka-lite crowdin .po files.

There must be a better way to determine if it's a khan academy .po or not.

File: centralserver/i18n/management/commands/update_language_packs.py [hunk 4]
```diff
@@ -558,15 +558,22 @@ def produce_outputs(src_po_files, dest_path, lang_code):
...
+                        logging.info("Found file %s", os.path.basename(src_file))
+                        src_po = polib.pofile(src_file)
+                        build_po.merge(src_po)
+                    else:
+                        logging.info("Ignoring %s because it's NOT for version %s or the .po files are not existed" %
```
Comment (cpauya) [inline centralserver/i18n/management/commands/update_language_packs.py]: This will never get executed as per comment above.

**Comment by cpauya** [comment]
@arceduardvincent You need to check for at least four (4) possible values for the `src_file` variable inside the loop:

(assuming all inputs are .po files)

1. A filename with the "KA Lite UI" and language code.

    > /var/folders/40/5c3y6nhs7tv00vv1nq5g0nvm0000gn/T/tmpDsmm4f/KA Lite UI/kalite-de.po

1. A filename with the same version as `SHORTVERSION` (KA-Lite)

    > /var/folders/40/5c3y6nhs7tv00vv1nq5g0nvm0000gn/T/tmpDsmm4f/versioned/0.13-django.po

1. A filename with a different version (KA-Lite)

    > /var/folders/40/5c3y6nhs7tv00vv1nq5g0nvm0000gn/T/tmpDsmm4f/versioned/0.12-django.po

1. A filename with no indicator for a version (Khan Academy .po file)

    > /var/folders/40/5c3y6nhs7tv00vv1nq5g0nvm0000gn/T/tmp2LH_QJ/1_high_priority_platform/_other_-de.po

Given the `SHORTVERSION` and language code, you should ignore the `.po` files that do NOT belong to the same version and concatenate everything else.

**Comment by arceduardvincent** [comment]
Hi @cpauya .

My solution in the case scenario that you have given is that when:
> when  the "KA Lite UI" and language code build po file.
> when the filename has a `SHORTVERSION ` build the .po file.
> when the filename has a (language code).po we build the .po file.
> else the file that has not meet this argument we pass and not build.

**Comment by arceduardvincent** [comment]
Hi @cpauya and @aronasorman .

I have fix some issue regarding the build of `.mo` files that does not include the approve crowdin tranlastion for the `perseus exercises` as suggested by @aronasorman .

This image below represent that the translation existed in the `.mo` files.
screen shot 2015-02-18 at 2 55 37 pm

The approve translation in the crowdin reffrence https://crowdin.com/translate/khanacademy/27617/enus-de#2973008 under the `learn.math.early-math.exercises.pot 
screen shot 2015-02-18 at 3 04 58 pm

**Comment by jamalex** [comment]
Is this good to go, @cpauya and @aronasorman?

**Comment by aronasorman** [comment]
Will review today.

**Comment by cpauya** [comment]
Yes @jamalex - I was able to replicate the translations on central repo with these steps:

1. on this branch, run `python manage.py update_language_packs -l de --no-srts --no-dubbed`
1. extract the `centralserver/media/language_packs/0.12/de.zip`
1. copy the `LC_MESSAGES` folder inside the `/locale/de/` folder
1. For my sample, I took the KA Crowdin string/s from [Official Khan Academy translations German translation](https://crowdin.com/project/khanacademy/de) - [learn.math.algebra-basics.exercises.pot - 43%](https://crowdin.com/translate/khanacademy/30006/enus-de).
1. run `python manage.py shell` and execute these:

```python
from django.utils.translation import activate, ugettext as _

activate("de")
s = "What is <code>X</code>?"
_(s)
```

Here's the output:

```python
>>> from django.utils.translation import activate, ugettext as _
>>>
>>> activate("de")
>>> s = "What is <code>X</code>?"
>>> _(s)
'Was ist <code>X</code>?'
>>>
```

The approved translations for Khan Academy's Crowdin account are now included in the generated `django.mo` file.  I don't know why we had the filters that @arceduardvincent commented-out, perhaps @aronasorman can verify too?

**Note**

I noticed that the new approved translations are not included on the `.zip` file when I use `python manage.py update_language_packs` as per above.  I did this by exctracting the `_crowdin_cache/ka-de.zip` and looked at the `ka-de/2_high_priority_content/learn.math.algebra-basics.exercises-de.po` file.

But when I downloaded directly from [the CrowdIn site](https://crowdin.com/project/khanacademy/de) - ie: `learn.math.algebra-basics.exercises.pot` - the downloaded `learn.math.algebra-basics.exercises-de.po` file contains the most recent translations.

Any ideas?  (perhaps am just sleepy or maybe missing something else)

File: centralserver/i18n/management/commands/update_language_packs.py [hunk 5]
```diff
@@ -497,37 +497,40 @@ def build_new_po(lang_code, src_path, dest_path=None, combine_with_po_file=None,
...
-                    except IOError: # either a parse error from polib, or file doesnt exist
-                        # TODO (ARON): capture all po files that return this error, show it to user
-                        continue
+        # Note: The filter below was commented due to some circumstance that the approve crowdin tranlastion for the
+        #       perseus exercises  was not included in the .mo file
 
-        else:
-            for po_file in src_po_files:
-                yield po_file
+        # remove all exercise po that is not about math
```
Comment (aronasorman) [inline centralserver/i18n/management/commands/update_language_packs.py]: remove this commented out code.

File: centralserver/i18n/management/commands/update_language_packs.py [hunk 6]
```diff
@@ -558,9 +561,13 @@ def produce_outputs(src_po_files, dest_path, lang_code):
                 js_po_file.save_as_mofile(js_mo_file)
             else:
                 # Make sure we only concatenate .po files of the same version that we need.
-                versioned_po_filename = os.path.join("versioned", "%s-django") % (version,)
-                kalite_po_filename = os.path.join("KA Lite UI", "kalite-%s.po") % (lang_code,)
-                if versioned_po_filename in src_file or kalite_po_filename in src_file:
+                ka_po_file = ""
```
Comment (aronasorman) [inline centralserver/i18n/management/commands/update_language_packs.py]: Why is this line necessary?

Comment (aronasorman) [inline centralserver/i18n/management/commands/update_language_packs.py]: Oh, because it might not be set in line 568.

Comment (aronasorman) [inline centralserver/i18n/management/commands/update_language_packs.py]: Can you just squash it into a single if statement then, so that it's clearer?

**Comment by arceduardvincent** [comment]
Hi @aronasorman .

I update the fix as you suggested.
Could you please take a look on it and verify it.

**Comment by arceduardvincent** [comment]
Ohh my bad sorry. I just misunderstand what @aronasorman said about the `if statement' 
I revert my fix and update it. 

Could you take review on this @jesumer?

**Comment by aronasorman** [comment]
Handling this one now. I'm also fixing the tests in this PR. Soon!!!

**Comment by arceduardvincent** [comment]
:+1:

**Comment by aronasorman** [comment]
It's working on the distributed side. Can confirm that the strings are in there. 

Merging it now so other people can manually test this. Will fix the remaining tests later.
